src/core/ext/transport/binder/utils/transport_stream_receiver_impl.cc

```diff
@@ -174,67 +178,67 @@ void TransportStreamReceiverImpl::NotifyRecvTrailingMetadata(
   cb(std::move(trailing_metadata), status);
 }
 
-void TransportStreamReceiverImpl::
-    CancelRecvMessageCallbacksDueToTrailingMetadata(StreamIdentifier id) {
+void TransportStreamReceiverImpl::OnRecvTrailingMetadata(StreamIdentifier id) {
   gpr_log(GPR_INFO, "%s id = %d is_client = %d", __func__, id, is_client_);
-  MessageDataCallbackType cb = nullptr;
   {
-    grpc_core::MutexLock l(&m_);
-    auto iter = message_cbs_.find(id);
-    if (iter != message_cbs_.end()) {
-      cb = std::move(iter->second);
-      message_cbs_.erase(iter);
+    InitialMetadataCallbackType cb = nullptr;
+    {
+      grpc_core::MutexLock l(&m_);
+      auto iter = initial_metadata_cbs_.find(id);
+      if (iter != initial_metadata_cbs_.end()) {
+        cb = std::move(iter->second);
+        initial_metadata_cbs_.erase(iter);
+      }
+      trailing_metadata_recvd_.insert(id);
+    }
+    if (cb != nullptr) {
+      // The registered callback will never be satisfied. Cancel it.
+      cb(absl::CancelledError(""));
     }
```

Comment:
Probably want a separate member function for readability

Reply:
Done.